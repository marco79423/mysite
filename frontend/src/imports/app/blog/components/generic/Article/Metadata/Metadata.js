import React from 'react'
import PropTypes from 'prop-types'
import ImmutablePropTypes from 'react-immutable-proptypes'
import NormalLink from '../../NormalLink'
import styled from 'styled-components'

import dateformat from 'dateformat'

const ItemGroup = styled.ul`
  font-size: 0.9rem;
  line-height: 1.6rem;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
`

const Item = styled.li`
  display: inline;
  &:not(:first-child) {
    margin-left: 8px;
  }
`

const Categories = styled.ul`
  display: inline;
`

const Category = styled.li`
  margin: 0;
  display: inline;
  &:not(:first-child):before {
    content: "/";
    margin: 0 3px;
  }
`

const ChickenImage = styled.img`
  margin-bottom: -5px;

  &:hover {
    transform: rotate(-10deg);
  }
`

const Tooltip = styled.span`
  display: ${props => props.hover ? 'inline-block' : 'none'};
  position: absolute;
  line-height: 1.2em;
  padding: 4px;
  border-radius: 3px;
  background: ${props => props.theme.page.main.content.article.metadata.tip.background};
  color: ${props => props.theme.page.main.content.article.metadata.tip.color};
  
  transition: opacity .3s ease-out;
  
  &:before {
    content: "";
    position: absolute;
    width: 0;
    height: 0;
    left: -15px;
    top: 4px;
    border: 8px solid transparent;
    border-right-color: ${props => props.theme.page.main.content.article.metadata.tip.background};
  }
  
  @media (max-width: 800px) {
    display: none;
  }
`

export default class Metadata extends React.PureComponent {
  static PropTypes = {
    categories: ImmutablePropTypes.listOf(ImmutablePropTypes.contains({
      slug: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired
    })).isRequired,
    chickenCount: PropTypes.number.isRequired,
    date: PropTypes.instanceOf(Date).isRequired,
    modifiedDate: PropTypes.instanceOf(Date)
  }

  state = {
    hoverChickenImage: false
  }

  renderCategoryItem = (categories) => {
    return (
      <Item>分類：
        <Categories>
          {categories.map(category => (
            <Category key={category.get('slug')}>
              <NormalLink to={`/articles/category/${category.get('slug')}/`}>{category.get('name')}</NormalLink>
            </Category>
          ))}
        </Categories>
      </Item>
    )
  }

  onMouseEnter = () => {
    this.setState({hoverChickenImage: true})
  }

  onMouseLeave = () => {
    this.setState({hoverChickenImage: false})
  }

  renderChickenCountItem = (chickenCount) => {
    const tooltip = '雞數：計算文長的常見計量單位，一般而言數字大小與文章長度呈正相關'
    return (
      <Item>字數：<ChickenImage src={require('./img/chicken.png')} alt='雞' onMouseEnter={this.onMouseEnter} onMouseLeave={this.onMouseLeave}/> x {chickenCount}<Tooltip hover={this.state.hoverChickenImage}>{tooltip}</Tooltip></Item>
    )
  }

  renderReleaseDateItem = (date) => {
    return (
      <Item>發表時間：{dateformat(date, 'yyyy/mm/dd')}</Item>
    )
  }

  renderModifiedDateItem = (modifiedDate) => {
    return (
      <Item>最後更新：{dateformat(modifiedDate, 'yyyy/mm/dd')}</Item>
    )
  }

  render () {
    return (
      <ItemGroup>
        {this.renderCategoryItem(this.props.categories)}
        {this.renderChickenCountItem(this.props.chickenCount)}
        {this.renderReleaseDateItem(this.props.date)}
        {this.props.modifiedDate && this.renderModifiedDateItem(this.props.modifiedDate)}
      </ItemGroup>
    )
  }
}

