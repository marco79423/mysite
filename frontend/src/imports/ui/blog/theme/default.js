const colorTable = {
  'cyan-50': '#E0F7FA',
  'cyan-600': '#00ACC1',
  'cyan-700': '#0097A7',
  'cyan-800': '#00838F',
  'cyan-900': '#006064',
  'cyan-A700': '#00B8D4',

  'teal-50':'#E0F2F1',
  'teal-100':'#B2DFDB',
  'teal-200':'#80CBC4',
  'teal-300':'#4DB6AC',
  'teal-400':'#26A69A',
  'teal-500':'#009688',
  'teal-600':'#00897B',
  'teal-700':'#00796B',
  'teal-800':'#00695C',
  'teal-900':'#004D40',
  'teal-A700':'#00BFA5',

  'grey-50': '#FAFAFA',
  'grey-200': '#EEEEEE',
  'grey-300': '#E0E0E0',
  'grey-600': '#757575',
  'grey-700': '#616161',
  'grey-900': '#212121'
}

export default {
  global: {
    fontColor: colorTable['grey-900'],
    link: {
      color: colorTable['cyan-700'],
      hoverColor: colorTable['cyan-900']
    }
  },
  page: {
    background: colorTable['grey-700'],
    header: {
      background: colorTable['teal-400'],
      titleColor: colorTable['grey-50']
    },
    nav: {
      background: colorTable['teal-600'],
      menu: {
        color: '#ECECEC',
        background: colorTable['teal-600'],
        hoverBackground: colorTable['cyan-900'],
        dropdown: {
          background: colorTable['teal-600'],
          borderColor: colorTable['teal-800']
        }
      }
    },
    main: {
      background: colorTable['grey-200'],
      content: {
        article: {
          background: 'white',
          metadata: {
            tip: {
              color: '#FFFFFF',
              background: colorTable['teal-600']
            }
          },
          header: {
            color: colorTable['grey-700']
          },
          block: {
            color: 'white',
            background: colorTable['teal-600'],
            borderColor: '#ccc'
          },
          doctestBlock: {
            color: 'white',
            background: '#6D6D6D',
          },
          note: {
            background: '#ECECEC',
            titleColor: colorTable['teal-800'],
            borderColor: '#ccc'
          },
          table: {
            header: {
              background: colorTable['teal-600'],
            }
          }
        },
        pagination: {
          color: colorTable['grey-200'],
          background: colorTable['teal-700'],
          hoverBackground: colorTable['teal-900'],
          activeBackground: colorTable['teal-900'],
        }
      },
      aside: {
        background: colorTable['grey-200'],
        darkerBackground: colorTable['grey-600'],

        section: {
          titleColor: colorTable['grey-700'],
          borderColor: colorTable['grey-200']
        }
      }
    },
    footer: {
      background: colorTable['teal-600'],
      copyrightColor: '#ECECEC',
    }
  }
}