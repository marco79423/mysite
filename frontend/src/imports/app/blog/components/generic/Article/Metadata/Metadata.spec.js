import Immutable from 'immutable'
import React from 'react'
import renderer from 'react-test-renderer'
import 'jest-styled-components'
import {ThemeProvider} from 'styled-components'

import Metadata from './Metadata'
import theme from '../../../../theme/default'

test('It should render Metadata component correctly', () => {
  const categories = Immutable.fromJS([
    {
      slug: 'slug',
      name: 'name'
    },
    {
      slug: 'slug2',
      name: 'name2'
    }
  ])
  const date = new Date(2017, 6, 9)
  const chickenCount = 1

  const component = renderer.create(
    <ThemeProvider theme={theme}>
      <Metadata categories={categories} chickenCount={chickenCount} date={date}/>
    </ThemeProvider>
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('It should render Metadata component correctly when target article has been modified', () => {
  const categories = Immutable.fromJS([{
    slug: 'slug',
    name: 'name'
  }])
  const chickenCount = 1
  const date = new Date(2017, 6, 9)
  const modifiedDate = new Date(2017, 10, 9)

  const component = renderer.create(
    <ThemeProvider theme={theme}>
      <Metadata categories={categories} chickenCount={chickenCount} date={date} modifiedDate={modifiedDate}/>
    </ThemeProvider>
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
