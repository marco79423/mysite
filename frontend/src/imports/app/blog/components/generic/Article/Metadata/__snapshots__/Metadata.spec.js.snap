// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`It should render Metadata component correctly 1`] = `
.c4 {
  color: #0097A7;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c4:hover {
  color: #006064;
}

.c0 {
  font-size: 0.9rem;
  line-height: 1.6rem;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.c1 {
  display: inline;
}

.c1:not(:first-child) {
  margin-left: 8px;
}

.c2 {
  display: inline;
}

.c3 {
  margin: 0;
  display: inline;
}

.c3:not(:first-child):before {
  content: "/";
  margin: 0 3px;
}

.c5 {
  margin-bottom: -5px;
}

.c5:hover {
  -webkit-transform: rotate(-10deg);
  -ms-transform: rotate(-10deg);
  transform: rotate(-10deg);
}

.c6 {
  display: none;
  position: absolute;
  line-height: 1.2em;
  padding: 4px;
  border-radius: 3px;
  background: #00897B;
  color: #FFFFFF;
  -webkit-transition: opacity .3s ease-out;
  transition: opacity .3s ease-out;
}

.c6:before {
  content: "";
  position: absolute;
  width: 0;
  height: 0;
  left: -15px;
  top: 4px;
  border: 8px solid transparent;
  border-right-color: #00897B;
}

@media (max-width:800px) {
  .c6 {
    display: none;
  }
}

<ul
  className="c0"
>
  <li
    className="c1"
  >
    分類：
    <ul
      className="c2"
    >
      <li
        className="c3"
      >
        <a
          className="c4"
          onClick={[Function]}
          style={Object {}}
        >
          name
        </a>
      </li>
      <li
        className="c3"
      >
        <a
          className="c4"
          onClick={[Function]}
          style={Object {}}
        >
          name2
        </a>
      </li>
    </ul>
  </li>
  <li
    className="c1"
  >
    字數：
    <img
      alt="雞"
      className="c5"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      src={
        Object {
          "default": Object {},
        }
      }
    />
     x 
    1
    <span
      className="c6"
    >
      雞數：計算文長的常見計量單位，一般而言數字大小與文章長度呈正相關
    </span>
  </li>
  <li
    className="c1"
  >
    發表時間：
    2017/07/09
  </li>
</ul>
`;

exports[`It should render Metadata component correctly when target article has been modified 1`] = `
.c4 {
  color: #0097A7;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c4:hover {
  color: #006064;
}

.c0 {
  font-size: 0.9rem;
  line-height: 1.6rem;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.c1 {
  display: inline;
}

.c1:not(:first-child) {
  margin-left: 8px;
}

.c2 {
  display: inline;
}

.c3 {
  margin: 0;
  display: inline;
}

.c3:not(:first-child):before {
  content: "/";
  margin: 0 3px;
}

.c5 {
  margin-bottom: -5px;
}

.c5:hover {
  -webkit-transform: rotate(-10deg);
  -ms-transform: rotate(-10deg);
  transform: rotate(-10deg);
}

.c6 {
  display: none;
  position: absolute;
  line-height: 1.2em;
  padding: 4px;
  border-radius: 3px;
  background: #00897B;
  color: #FFFFFF;
  -webkit-transition: opacity .3s ease-out;
  transition: opacity .3s ease-out;
}

.c6:before {
  content: "";
  position: absolute;
  width: 0;
  height: 0;
  left: -15px;
  top: 4px;
  border: 8px solid transparent;
  border-right-color: #00897B;
}

@media (max-width:800px) {
  .c6 {
    display: none;
  }
}

<ul
  className="c0"
>
  <li
    className="c1"
  >
    分類：
    <ul
      className="c2"
    >
      <li
        className="c3"
      >
        <a
          className="c4"
          onClick={[Function]}
          style={Object {}}
        >
          name
        </a>
      </li>
    </ul>
  </li>
  <li
    className="c1"
  >
    字數：
    <img
      alt="雞"
      className="c5"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      src={
        Object {
          "default": Object {},
        }
      }
    />
     x 
    1
    <span
      className="c6"
    >
      雞數：計算文長的常見計量單位，一般而言數字大小與文章長度呈正相關
    </span>
  </li>
  <li
    className="c1"
  >
    發表時間：
    2017/07/09
  </li>
  <li
    className="c1"
  >
    最後更新：
    2017/11/09
  </li>
</ul>
`;
